File: calculator.py
from cmath import sqrt

from flask import Flask, render_template, request


app = Flask(__name__, template_folder=".", static_folder=".")


def format_number(value):
    if abs(value) < 1e-10:
        value = 0

    if isinstance(value, float) and value.is_integer():
        return str(int(value))

    return f"{value:.4g}"


def format_root(root):
    real = root.real
    imaginary = root.imag

    if abs(imaginary) < 1e-10:
        return format_number(real)

    sign = "+" if imaginary >= 0 else "-"
    return f"{format_number(real)} {sign} {format_number(abs(imaginary))}i"


def find_quadratic_zeros(a, b, c):
    if a == 0:
        return "a cannot be 0 for a quadratic equation."

    discriminant = b**2 - 4 * a * c
    first_zero = (-b + sqrt(discriminant)) / (2 * a)
    second_zero = (-b - sqrt(discriminant)) / (2 * a)

    return f"x = {format_root(first_zero)} and x = {format_root(second_zero)}"


@app.route("/")
def home():
    return render_template(
        "homepage.html",
        result="Your zeros will appear here after calculation.",
        values={"a": "", "b": "", "c": ""},
    )


@app.route("/calculate", methods=["POST"])
def calculate():
    values = {
        "a": request.form.get("a", ""),
        "b": request.form.get("b", ""),
        "c": request.form.get("c", ""),
    }

    try:
        a = float(values["a"])
        b = float(values["b"])
        c = float(values["c"])
        result = find_quadratic_zeros(a, b, c)
    except ValueError:
        result = "Please enter valid numbers for a, b, and c."

    return render_template("homepage.html", result=result, values=values)


if __name__ == "__main__":
    app.run(debug=True)

File: calculator.js
const form = document.querySelector("#calculator-form");
const result = document.querySelector("#result");

function formatNumber(value) {
  if (Math.abs(value) < 1e-10) {
    return "0";
  }

  if (Number.isInteger(value)) {
    return String(value);
  }

  return Number(value.toFixed(4)).toString();
}

function formatRoot(real, imaginary = 0) {
  if (Math.abs(imaginary) < 1e-10) {
    return formatNumber(real);
  }

  const sign = imaginary >= 0 ? "+" : "-";
  return `${formatNumber(real)} ${sign} ${formatNumber(Math.abs(imaginary))}i`;
}

function findQuadraticZeros(a, b, c) {
  if (a === 0) {
    return "a cannot be 0 for a quadratic equation.";
  }

  const discriminant = b ** 2 - 4 * a * c;
  const denominator = 2 * a;

  if (discriminant >= 0) {
    const firstZero = (-b + Math.sqrt(discriminant)) / denominator;
    const secondZero = (-b - Math.sqrt(discriminant)) / denominator;

    return `x = ${formatRoot(firstZero)} and x = ${formatRoot(secondZero)}`;
  }

  const real = -b / denominator;
  const imaginary = Math.sqrt(Math.abs(discriminant)) / Math.abs(denominator);

  return `x = ${formatRoot(real, imaginary)} and x = ${formatRoot(real, -imaginary)}`;
}

form.addEventListener("submit", (event) => {
  event.preventDefault();

  const formData = new FormData(form);
  const a = Number(formData.get("a"));
  const b = Number(formData.get("b"));
  const c = Number(formData.get("c"));

  if (!Number.isFinite(a) || !Number.isFinite(b) || !Number.isFinite(c)) {
    result.textContent = "Please enter valid numbers for a, b, and c.";
    return;
  }

  result.textContent = findQuadraticZeros(a, b, c);
});
